File: docs/truth_assurance_pipeline/experiments/tap_e7_artifact_assurance/tap-e7-base-implementation-b/tools/recompute_config_fingerprint.js
'use strict';
const fs = require('fs'), path = require('path'), crypto = require('crypto');
function canon(o){ if(o===null||typeof o!=='object')return JSON.stringify(o);
  if(Array.isArray(o))return '['+o.map(canon).join(',')+']';
  return '{'+Object.keys(o).sort().map(k=>JSON.stringify(k)+':'+canon(o[k])).join(',')+'}'; }
const sha=(b)=>'sha-256:'+crypto.createHash('sha256').update(b).digest('hex');
function recompute(pkg){
  const rel=JSON.parse(fs.readFileSync(path.join(pkg,'manifest/release-manifest.json')));
  const rm=JSON.parse(fs.readFileSync(path.join(pkg,'manifest/resource-manifest.json')));
  const runtime=rm.resources.filter(e=>e.outcome_affecting);
  const obj={target_spec:rel.target_specification,target_profile:rel.target_profile,
    canonicalization:rel.canonicalization,thresholds:{T_accept:0.85,T_reject:0.35},
    runtime_resources:runtime.map(e=>({path:e.path,sha256:e.sha256}))};
  return [sha(Buffer.from(canon(obj)+'\n','utf8')), rel.roots.config_fingerprint, runtime];
}
module.exports={recompute,canon,sha};
if(require.main===module){ const [g,w,rt]=recompute(process.argv[2]);
  console.log('recomputed:',g); console.log('package   :',w);
  console.log('runtime_resources:',rt.length,'| corpus excluded:',!rt.some(e=>e.path.startsWith('corpus/')));
  console.log(g===w?'MATCH':'MISMATCH'); process.exit(g===w?0:2); }

File: docs/truth_assurance_pipeline/experiments/tap_e7_artifact_assurance/tap-e7-base-implementation-a/tools/recompute_config_fingerprint.py
#!/usr/bin/env python3
"""Independently recompute the runtime config fingerprint from published package data.
Recipe is reconstructed from the release-manifest (target ids, canonicalization, note that
corpus_root is excluded), the frozen thresholds, and the resource-manifest outcome_affecting
flags. If our computed value differs from the package value, execution must stop."""
import hashlib, json, os, sys
def canon(o): return json.dumps(o, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
def sha(b): return "sha-256:" + hashlib.sha256(b).hexdigest()

def recompute(pkg):
    rel = json.load(open(os.path.join(pkg, "manifest/release-manifest.json")))
    rm = json.load(open(os.path.join(pkg, "manifest/resource-manifest.json")))
    runtime = [e for e in rm["resources"] if e.get("outcome_affecting")]
    obj = {"target_spec": rel["target_specification"],
           "target_profile": rel["target_profile"],
           "canonicalization": rel["canonicalization"],
           "thresholds": {"T_accept": 0.85, "T_reject": 0.35},
           "runtime_resources": [{"path": e["path"], "sha256": e["sha256"]} for e in runtime]}
    return sha((canon(obj) + "\n").encode()), rel["roots"]["config_fingerprint"]

if __name__ == "__main__":
    pkg = sys.argv[1]
    got, want = recompute(pkg)
    print("recomputed:", got)
    print("package   :", want)
    print("MATCH" if got == want else "MISMATCH")
    sys.exit(0 if got == want else 2)
